src/Trains.NET.Rendering/LayerRenderer/TerrainRenderer.cs

```diff
@@ -0,0 +1,157 @@
+            foreach (var terrain in _terrainLayout)
+            {
+                var contourLevel = CalculateContourLevel(terrain.Altitude);
+                var contourPoints = topography.ContainsKey(contourLevel) ? topography[contourLevel] : new List<ContourPoint>();
+
+                var adjacentTerrainLookups = new List<Func<Terrain, Terrain>>
+                {
+                    _terrainLayout.GetAdjacentTerrainUp,
+                    _terrainLayout.GetAdjacentTerrainDown,
+                    _terrainLayout.GetAdjacentTerrainLeft,
+                    _terrainLayout.GetAdjacentTerrainRight,
+                };
+
+                foreach (var adjacentTerrainLookup in adjacentTerrainLookups)
+                {
+                    var adjacentContourPoint = CalculateBorderingContourPointIfAdjacentTerrainIsLower(adjacentTerrainLookup, terrain, contourLevel);
+                    if (adjacentContourPoint != null && !contourPoints.Contains(adjacentContourPoint.Value))
+                        contourPoints.Add(adjacentContourPoint.Value);
+                }
+
+                topography[contourLevel] = contourPoints;
+            }
+
+            return topography;
+        }
+
+        private ContourPoint? CalculateBorderingContourPointIfAdjacentTerrainIsLower(Func<Terrain, Terrain> getAdjacentTerrain, Terrain terrain, int contourLevel)
+        {
+            var adjacentTerrain = getAdjacentTerrain(terrain);
+
+            var adjacentContourLevel = CalculateContourLevel(adjacentTerrain.Altitude);
+
+            if (adjacentContourLevel >= contourLevel) return null;
+
+           return FindContourPointBetweenAdjacentTerrain(terrain, adjacentTerrain);
+        }
+
+        private ContourPoint FindContourPointBetweenAdjacentTerrain(Terrain sourceTerrain, Terrain adjacentTerrain)
+        {
+            var columnDelta = sourceTerrain.Column - adjacentTerrain.Column;
+            var rowDelta = sourceTerrain.Row - adjacentTerrain.Row;
+
+            // Adjacent means that there should only be exactly one difference between either Columns or Rows (And the other should have no difference)
+            if ((Math.Abs(columnDelta) + Math.Abs(rowDelta)) != 1)
+                throw new ArgumentException("Trying to find contour point between terrain that is not adjacent");
+
+            if (columnDelta != 0)
+            {
+                // Different columns, must be same row
+                return new ContourPoint
+                {
+                    // columnDelta gt zero means we have source on right, adjacent on left so centre point is between them.. which means source edge (since cell edges are on left)
+                    // columnDelta lt zero means we have source on left, adjacent on right so centre point is again between them.. which means adjacent edge 
+                    Column = (columnDelta > 0 ? sourceTerrain.Column : adjacentTerrain.Column),
+                    Row = (sourceTerrain.Row + 0.5f),
```

Comment:
Yeah, so I would leave Column and Row as int, and just add `_trackParameters.CellSize / 2` to get to the middle of the cell.

Reply:
As discussed, I think this needs more thought beyond the scope of this PR.

Reply:
Unresolving to work on discussed change

Reply:
Done